pkg/app/appconf/store.go

```diff
@@ -0,0 +1,298 @@
+package appconf
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/google/renameio/v2/maybe"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	"golang.org/x/exp/slices"
+
+	"github.com/vanti-dev/sc-bos/internal/confmerge"
+	"github.com/vanti-dev/sc-bos/pkg/auto"
+	"github.com/vanti-dev/sc-bos/pkg/block"
+	"github.com/vanti-dev/sc-bos/pkg/driver"
+	"github.com/vanti-dev/sc-bos/pkg/zone"
+)
+
+type Store struct {
+	logger  *zap.Logger
+	backing confmerge.Store
+
+	m      sync.Mutex
+	active Config
+}
+
+func LoadStore(external Config, schema Schema, storeDir string, logger *zap.Logger) (*Store, error) {
+	store := confmerge.NewDirStore(storeDir)
+	active, patches, err := confmerge.Merge(external, store, schema.Blocks())
+	err = multierr.Append(err, saveConfigPatches(patches, filepath.Join(storeDir, patchDirName), logger))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Store{
+		logger:  logger,
+		backing: store,
+		active:  active,
+	}, nil
+}
+
+func (s *Store) Active() Config {
+	s.m.Lock()
+	defer s.m.Unlock()
+	return s.active.clone()
+}
+
+func (s *Store) Drivers() *DriverStore {
+	return &DriverStore{store: s}
+}
+
+func (s *Store) Automations() *AutomationStore {
+	return &AutomationStore{store: s}
+}
+
+func (s *Store) Zones() *ZoneStore {
+	return &ZoneStore{store: s}
+}
+
+func (s *Store) save(updated Config) error {
+	encoded, err := json.MarshalIndent(updated, "", "  ")
+	if err != nil {
+		return err
+	}
+	err = s.backing.SetActiveConfig(encoded)
+	if err != nil {
+		return err
+	}
+	s.active = updated
+	return nil
+}
+
+type DriverStore struct {
+	store *Store
+}
+
+func (ds *DriverStore) SaveConfig(_ context.Context, name string, typ string, data []byte) error {
+	ds.store.m.Lock()
+	defer ds.store.m.Unlock()
+
+	if name == "" {
+		return errors.New("driver name is required")
+	}
+
+	// insert the updated driver into a copy of the config
+	updated := ds.store.active.clone()
+	idx := slices.IndexFunc(updated.Drivers, func(c driver.RawConfig) bool {
+		return c.Name == name
+	})
+	var driverCfg driver.RawConfig
+	if idx < 0 {
+		// add new driver
+		if typ == "" {
+			return errors.New("driver type is required to add a new driver")
+		}
+		driverCfg = driver.RawConfig{
+			BaseConfig: driver.BaseConfig{
+				Name:     name,
+				Type:     typ,
+				Disabled: false,
+			},
+			Raw: data,
+		}
+		updated.Drivers = append(updated.Drivers, driverCfg)
+	} else {
+		existing := updated.Drivers[idx]
+		if typ != "" && typ != existing.Type {
+			return errors.New("driver type cannot be changed")
+		}
+		driverCfg = existing
+		driverCfg.Raw = data
+		updated.Drivers[idx] = driverCfg
+	}
+
+	// test that the updated config marshalls successfully - we don't want to store a bad config
+	_, err := json.Marshal(updated)
```

Comment:
Could we get away with just testing the new/updated driver config marshals? I'm afraid that if we add bulk update features later this json.Marshal will cause it to be O(n!) complexity

Reply:
Yes, good idea. Done.